os/drivers/usbhost/uvc/usbhost_uvc_ctrl.c

```diff
@@ -0,0 +1,1357 @@
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+int uvc_ctrl_restore_values(uvc_state_t *dev)
+{
+	struct uvc_control *ctrl;
+	struct uvc_entity *entity = queue_entry(dev->entities.head, struct uvc_entity, list);
+	unsigned int i;
+	int ret = OK;
+
+	/* Walk the entities list and restore controls when possible. */
+	for (; &entity->list != (dq_entry_t *)&dev->entities; entity = queue_entry(entity->list.flink, struct uvc_entity, list)) {
+
+		for (i = 0; i < entity->ncontrols; ++i) {
+			ctrl = &entity->controls[i];
+
+			if (!CTRL_IS_SUPPORTED(ctrl->status) || !CTRL_IS_MODIFIED(ctrl->status) || (ctrl->info.flags & F_RESTORE) == 0) {
+				continue;
+			}
+
+			uvdbg("restoring control %pUl/%u/%u\n", ctrl->info.entity, ctrl->info.index, ctrl->info.selector);
+			CTRL_SET_DIRTY(ctrl->status);
+
+			if (ctrl->info.flags & F_AUTO || !(ctrl->info.flags & F_GET_CUR)) {
+				CTRL_CLEAR_LOADED(ctrl->status);
+			}
+
+			ret = uvc_query_ctrl(dev, UVC_SET_CUR, ctrl->entity->id, dev->intfnum, ctrl->info.selector, ctrl->cur, ctrl->info.size);
+
+			if (ret < 0) {
+				memcpy(ctrl->cur, ctrl->bak, ctrl->info.size);
+			} else {
+				ret = OK;
+			}
+
+			CTRL_CLEAR_DIRTY(ctrl->status);
+		}
+	}
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: allocate_ctrl_vals
+ *
+ * Description:
+ *   This function allocate memory for min/max/default/current/backup.
+ *
+ *   This function is called from the uvc_ctrl_add_info method.
+ *
+ * Input Parameters:
+ *   ctrl - reference to the control object.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+static int allocate_ctrl_vals(struct uvc_control *ctrl)
+{
+	int ret = OK;
+	/* Allocate memory for each of the controls */
+	ctrl->min = kmm_zalloc(ctrl->info.size);
+	if (ctrl->min == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->max = kmm_zalloc(ctrl->info.size);
+	if (ctrl->max == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->cur = kmm_zalloc(ctrl->info.size);
+	if (ctrl->cur == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->step = kmm_zalloc(ctrl->info.size);
+	if (ctrl->step == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->def = kmm_zalloc(ctrl->info.size);
+	if (ctrl->def == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->bak = kmm_zalloc(ctrl->info.size);
+	if (ctrl->bak == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+done:
+	if (ret < 0) {
+		if (ctrl->min) {
+			kmm_free(ctrl->min);
+		}
+		if (ctrl->max) {
+			kmm_free(ctrl->max);
+		}
+		if (ctrl->cur) {
+			kmm_free(ctrl->cur);
+		}
+		if (ctrl->bak) {
+			kmm_free(ctrl->bak);
+		}
+		if (ctrl->step) {
+			kmm_free(ctrl->step);
+		}
+		if (ctrl->def) {
+			kmm_free(ctrl->def);
+		}
+	}
+	return ret;
```

Comment:
How about: -

```
zalloc 1
    return

zalloc 2
    goto err 2

zalloc 3
     goto err 3
..
..
..
zalloc n
    goto err n

return OK

err n:
    free n-1
..
..
..
err 3
    free 2

err 2
    free 1

return ret;
```

Reply:
Why to keep multiple return in first place, also multiple labels. It looks clean to me.

Reply:
Ohkay